### PR TITLE
SuJSON: disallow non-uints, improve tests and errors

### PR DESCRIPTION
**Pull request by oxade**
Only allow unsigned ints in arrays in SuiJSON homogenous check.

### CODE REVIEW

File: crates/sui-json/src/lib.rs [hunk 1]
```diff
@@ -293,29 +332,38 @@ fn is_homogeneous_rec(curr_q: &mut VecDeque<&JsonValue>) -> bool {
     // Process all in this queue/level
     while !curr_q.is_empty() {
         // Okay to unwrap since we know values exist
-        let curr = match curr_q.pop_front().unwrap() {
+        let v = curr_q.pop_front().unwrap();
+        let curr = match v {
             JsonValue::Bool(_) => ValidJsonType::Bool,
-            JsonValue::Number(_) => ValidJsonType::Number,
+            JsonValue::Number(x) if x.is_u64() => ValidJsonType::Number,
```
Comment (patrickkuo) [inline crates/sui-json/src/lib.rs:338]: we only support u64?

Comment (oxade) [reply crates/sui-json/src/lib.rs:338]: JSON bid width is u64. That's the only way to check if numbers are unsigned.

**Review by stella3d — APPROVED** [review]
Looks good.  Some comments could be clarified or expanded but the tests look thorough and the logic is good

File: crates/sui-json/src/lib.rs [hunk 2]
```diff
@@ -28,6 +28,47 @@ const HEX_PREFIX: &str = "0x";
 #[cfg(test)]
 mod tests;
 
+/// A list of error categories encountered when parsing numbers.
+#[derive(Debug, PartialEq, Eq, Clone, Copy, Hash)]
+pub enum SuiJsonValueErrorKind {
+    /// JSON value must be of specific types.
+    ValueTypeNotAllowed,
+
+    /// JSON arrays must be homogeneous.
```
[the remark this answers is not in the document]
Comment (oxade) [reply crates/sui-json/src/lib.rs:37]: Agreed.

File: crates/sui-json/src/tests.rs [hunk 3]
```diff
@@ -125,234 +184,215 @@ fn test_basic_args_linter_pure_args() {
...
-            )))))),
-            Some(
-                bcs::to_bytes(&vec![
-                    vec![vec![9u64, 53u64, 434u64], vec![0u64], vec![300u64]],
-                    vec![],
-                    vec![vec![332u64], vec![4u64, 5u64, 6u64, 7u64]],
-                ])
-                .unwrap(),
-            ),
+            MoveTypeLayout::Vector(Box::new(MoveTypeLayout::Vector(Box::new(
```
[the remark this answers is not in the document]
Comment (oxade) [reply crates/sui-json/src/tests.rs:375]: Right I'm not sure who added this but it was probably @awelc  or @tnowacki . Maybe they can explain.

[the remark this answers is not in the document]
Comment (awelc) [reply crates/sui-json/src/tests.rs:375]: I did not add this test, but I am also not sure what the question is about
- the nested `MoveTypeLayout`? As the comment says it's a nested (vector of vector of vector...) vector of u64s
- the BCS encoding? It simply replicates what's defined in JSON at the beginning of this particular test

**Pull request merged**